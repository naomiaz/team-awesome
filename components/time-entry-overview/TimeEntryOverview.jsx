import React from 'react';

import TimeEntryForm from '../time-entry-form/TimeEntryForm';
import TimeEntryDetail from '../time-entry-detail/TimeEntryDetail';
import timeEntryData from './mock-database.json';

import './time-entry-overview.scss';

// const convertDateToISO = (entryDate) => new Date(Date.parse(entryDate.value)).toISOString();
// const convertTime = (entryTime) => (Date.parse(entryTime.value))/1000/;


class TimeEntryOverview extends React.Component {
  state = {
    isFormVisible: false,
    timeEntries: timeEntryData,
    client: 'Default Client',
    activity: 'Default Activity',
    date: '01-01-2001',
    from: '00:01',
    to: '00:10'
  }

  onEntrySubmit = (event) => {
    event.preventDefault();
    const newTimeEntry = {
      projectDate: this.state.date,
      projectId: '001',
      client: this.state.client,
      activity: this.state.activity,
      timeFrom: this.state.from,
      timeTo: this.state.to
    };

    // console.log(this.state);

    this.setState(prevState => ({
      ...prevState,
      timeEntries: [
        newTimeEntry,
        ...prevState.timeEntries
      ]
    }));

    console.log(this.state);

    // console.log(newTimeEntry, ...timeEntries);
    // generateTimesheet(newTimeEntry);

    // projectDate: convertDate(document.querySelector('.time-entry__input#date').value),
    // projectId: parseInt(timeEntries[index-1], 10),
    // client: document.querySelector('.time-entry__input#client').textContent,
    // activity: document.querySelector('.time-entry__input#activity').textContent,
    // timeFrom: '10:00',
    // timeTo: '15:00'
  };

  formVisible = () => {
    this.setState(prevState => ({
      isFormVisible: !prevState.isFormVisible
    }));
  }

  handleChange = ({ target }) => {
    // handleChange vult de state met de waardes uit het formulier (setState(this.state))
    this.setState(prevState => ({
      ...prevState,
      [target.name]: target.value
    }));
  }

  render() {
    const { isFormVisible, timeEntries } = this.state;
    return (
      <React.Fragment>
        <TimeEntryForm
          formVisible={this.formVisible}
          handleChange={this.handleChange}
          isFormVisible={isFormVisible}
          onEntrySubmit={this.onEntrySubmit}
        />

        <section className="row time-entry-overview">
          {timeEntries.map((currentTimeEntry, index, array) => (
            // if (index === 0 ) { date + component } ------->> 0 is falsy
            // if (currentTimeEntry.date !== previousTimeEntry.date) { date + component }
            // if (currentTimeEntry.date === previousTimeEntry.date) { component }
            <React.Fragment key={currentTimeEntry.projectId}>
              {(!index || (currentTimeEntry.projectDate !== array[index - 1].projectDate)) && (
              <h3 className="text--secondary time-entry__date">
                { currentTimeEntry.projectDate }
              </h3>
              )}
              <TimeEntryDetail {...currentTimeEntry} />
            </React.Fragment>
          ))}
        </section>
      </React.Fragment>
    );
  }
}

export default TimeEntryOverview;
