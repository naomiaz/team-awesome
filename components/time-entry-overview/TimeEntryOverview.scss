@import "../../shared/styles/main";

.time-entry {
  &-overview {
    margin-top: 32px;
  }

  &__date {
    font-weight: 400;
    margin-top: 25px;

    .text--secondary {
      font-size: 24px;
      margin-bottom: 10px;
    }
  }

  &__project, {
    border: 1px solid $color-monochrome-6;
    border-left: 4px solid $color-primary;
    border-radius: $border-radius-primary;
    display: flex;
    flex-flow: row;
    justify-content: space-between;
    padding: 30px 16px;

    &__employer,
    &__time {
      flex-basis: 50%;
    }

    &__time {
      text-align: right;
    }
  }
}
