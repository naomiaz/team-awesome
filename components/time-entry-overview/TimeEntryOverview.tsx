import React from 'react';
import PageHeader from '../../shared/components/page-header/PageHeader';
import TimeEntryForm from '../time-entry-form/TimeEntryForm';
import TimeEntryItem from '../time-entry-item/TimeEntryItem';
import { TimeEntryModel } from '../../ducks/time-entries';
import { ClientNameModel } from '../../ducks/clients';
import { getRelativeDay, calculateDurationPerDay } from '../../services/date-time/date-time';

import './time-entry-overview.scss';

export interface TimeEntryOverviewProps {
  activeFilter: string;
  timeEntries: TimeEntryModel[];
  clientNames: ClientNameModel[];
  isFormSaving: boolean;
  isFormVisible: boolean;
  onDeleteTimeEntry: (id: number) => void;
  onFilterTimeEntries: (filterValue: string) => void;
  onRequestTimeEntries: () => void;
  onRequestClients: () => void;
  onSaveTimeEntry: (newTimeEntry: TimeEntryModel) => void;
  onToggleFormVisibility:(isFormVisible: boolean) => void;
  pageTitle: string;
  unitCount: number;
  unitSingular: string;
  unitPlural: string;
}


class TimeEntryOverview extends React.Component <TimeEntryOverviewProps> {
  componentDidMount() {
    this.props.onRequestClients();
    this.props.onRequestTimeEntries();
  }

  onEntrySubmit = (newTimeEntry) => {
    this.props.onSaveTimeEntry(newTimeEntry);
  };

  onEntryDelete = (id) => {
    this.props.onDeleteTimeEntry(id);
  };

  render(): React.ReactNode {
    const {
      activeFilter,
      clientNames,
      timeEntries,
      isFormSaving,
      isFormVisible,
      onFilterTimeEntries,
      onToggleFormVisibility
    } = this.props;
    const dateOptions = { weekday: 'long', day: 'numeric', month: '2-digit' };
    return (
      <React.Fragment>
        <PageHeader
          selectBox={[{
            onChange: (event) => onFilterTimeEntries(event.target.value),
            options: [{ title: 'All clients:', value: '' }, ...clientNames],
            selectedValue: activeFilter
          }]}
          pageTitle="Timesheets"
          unitCount={timeEntries.length}
          unitPlural="Entries"
          unitSingular="Entry"
        />

        <TimeEntryForm
          isFormSaving={isFormSaving}
          isFormVisible={isFormVisible}
          onToggleFormVisibility={onToggleFormVisibility}
          onEntrySubmit={this.onEntrySubmit}
          clientNames={clientNames}
        />

        <section className="time-entry-overview">
          {timeEntries.map((currentTimeEntry, index, array) => {
            // if (index === 0 ) { date + component } ------->> 0 is falsy
            // if (currentTimeEntry.date !== previousTimeEntry.date) { date + component }
            // if (currentTimeEntry.date === previousTimeEntry.date) { component }
            const dateFormatted = (date) => new Date(date).toLocaleDateString('en-NL', dateOptions).replace('/', '-').replace(',', '');
            return (
              <React.Fragment key={currentTimeEntry.id}>
                {(!index || (currentTimeEntry.date !== array[index - 1].date)) && (
                  <h2 className="time-entry__date-row">
                    <span>
                      {`${dateFormatted(currentTimeEntry.date)}
                      ${getRelativeDay(currentTimeEntry.timeFrom)}`}
                    </span>

                    <span>
                      {new Date(calculateDurationPerDay(timeEntries, currentTimeEntry.date))
                        .toLocaleTimeString('nl-NL', { hour: 'numeric', minute: 'numeric' })}
                    </span>
                  </h2>
                )}
                <TimeEntryItem
                  {...currentTimeEntry}
                  onEntryDelete={this.onEntryDelete}
                />
              </React.Fragment>
            );
          })
        }
        </section>
      </React.Fragment>
    );
  }
}

export default TimeEntryOverview;
