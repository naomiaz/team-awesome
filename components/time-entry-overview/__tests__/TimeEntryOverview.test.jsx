import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import TimeEntryOverview from '../TimeEntryOverview';

it('renders correctly', () => {
  const wrapper = shallow(
    <TimeEntryOverview
      isFormSaving={false}
      isFormVisible={false}
      onDeleteTimeEntry={() => {}}
      onRequestTimeEntries={() => {}}
      onSaveTimeEntry={() => {}}
      onToggleFormVisibility={() => {}}
      timeEntries={[{
        client: 'Mercedes',
        activity: 'Design',
        date: '2018-08-27',
        timeFrom: '2018-08-27T08:00:00.000Z',
        timeTo: '2018-08-27T12:00:00.000Z',
        id: 11
      }]}
    />
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});
