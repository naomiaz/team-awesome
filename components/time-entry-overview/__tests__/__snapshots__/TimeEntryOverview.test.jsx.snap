// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Fragment>
  <TimeEntryForm
    isFormSaving={false}
    isFormVisible={false}
    onEntrySubmit={[Function]}
    onToggleFormVisibility={[Function]}
    timeEntries={
      Array [
        Object {
          "activity": "Design",
          "client": "Mercedes",
          "date": "2018-08-27",
          "id": 11,
          "timeFrom": "2018-08-27T08:00:00.000Z",
          "timeTo": "2018-08-27T12:00:00.000Z",
        },
      ]
    }
  />
  <section
    className="row time-entry-overview"
  >
    <div
      key="11"
    >
      <div
        className="time-entry__date-row"
      >
        <span
          className="text--secondary"
        >
          Monday 08-27
                      
        </span>
        <span
          className="text--secondary"
        >
          04:00
        </span>
      </div>
      <TimeEntryDetail
        activity="Design"
        client="Mercedes"
        date="2018-08-27"
        id={11}
        onEntryDelete={[Function]}
        timeFrom="2018-08-27T08:00:00.000Z"
        timeTo="2018-08-27T12:00:00.000Z"
      />
    </div>
  </section>
</Fragment>
`;
